### PR TITLE
feat: make the linter coalesce the passed-in values before running values tests

### LINKED ISSUE #7756: Helm lint (v3) doesn't take values files into account during schema validation
Edit: Steps to reproduce: https://github.com/helm/helm/issues/7756#issuecomment-617889123

I have a bunch of charts with `values.schema.json` files that won't pass schema validation without passing in necessary, environment-specific values. I don't want these values in the base `values.yaml` file because they are environment-specific, but should fail at install/upgrade time.

When running `helm lint ${chart} --values ./lint-values.yaml` I'm getting errors like below **even though these values are provided in `./lint-values.yaml`**:
```
==> Linting ${chart}
[ERROR] values.yaml:
- ingress.hosts: Array must have at least 1 items
...

Error: 1 chart(s) linted, 1 chart(s) failed
```

Important to note that `helm template ${chart} --values ./line-values.yaml` does not throw any schema validation errors. When I run `helm template ${chart}` I see the same errors as above:
```
Error: values don't meet the specifications of the schema(s) in the following chart(s):
agent-services:
- ingress.hosts: Array must have at least 1 items
...
```

I think this is a bug. `helm lint` should run values schema validation against the final `.Values` object like the docs say:
> Note that the schema is applied to the final `.Values` object, and not just to the `values.yaml` file. This means that the following `yaml` file is valid, given that the chart is installed with the appropriate `--set` option shown below.

Source: https://helm.sh/docs/topics/charts/#schema-files

---

Output of `helm version`:
```
version.BuildInfo{Version:"v3.1.1", GitCommit:"afe70585407b420d0097d07b21c47dc511525ac8", GitTreeState:"clean", GoVersion:"go1.13.8"}
```

Output of `kubectl version`: not relevant

Cloud Provider/Platform (AKS, GKE, Minikube etc.): not relevant

### PR DESCRIPTION
**Pull request by technosophos**
When support for `helm lint --values` was added, this support did not extend to the values file tests. This PR extends the support to values files, which lets the user schema-test the combination of the chart and the passed-in values.

Also added tests for other values lints.

Closes #7756 

Signed-off-by: Matt Butcher <matt.butcher@microsoft.com>

### CODE REVIEW

File: pkg/lint/rules/values.go
Comment (sean-krail) [inline pkg/lint/rules/values.go:39]: `iis` -> `is`

**Review by sean-krail — CHANGES_REQUESTED** [review]
Looks good and thanks for pushing a fix so quickly. Just found a small spelling mistake.

File: pkg/lint/rules/values.go
Comment (technosophos) [reply pkg/lint/rules/values.go:39]: Good catch. Thanks.

**Review by sean-krail — APPROVED** [review]

**Review by mattfarina — APPROVED** [review]
lgtm

**Comment by technosophos** [comment]
I believe this needs one more maintainer review for merge.

**Review by bacongobbler — APPROVED** [review]

File: pkg/lint/lint.go
Comment (robbie-demuth) [inline pkg/lint/lint.go:33]: Are there no concerns about changing the internals of this function per https://github.com/helm/helm/pull/7417#discussion_r375961183 where I tried to make a similar change? While this function's signature isn't changing, it's output certainly might in cases where it was passed a non-empty values map, no?

Comment (robbie-demuth) [reply pkg/lint/lint.go:33]: (Don't get me wrong, I'm all for this change. I was just curious)

Comment (technosophos) [reply pkg/lint/lint.go:33]: The issue that was raised with your change (which I hadn't seen until yesterday) was that you changed the public signature of `Values()`. That is an API breaking change, and is not allowed until Helm 4. @bacongobbler was not blocking your PR based on an abstract notion of "sameness." He was blocking it based on our policy to not introduce breaking changes to the public API. It's true that we do want to make sure things continue to function similarly, but I think you may have gotten sidetracked by that. Fixing a bug is, of course, a proper behavior change. He was just making a sidelong comment that if no values were provided, it should still function the same.

FWIW, I don't have a preference over which our our PRs gets merged. So if you'd rather update yours, I'll review the changes. If I had noticed it before writing this one, I would have offered feedback then instead of doing what equates to duplicate work.

Comment (robbie-demuth) [reply pkg/lint/lint.go:33]: Thanks for the explanation! I'm fine with this moving forward since I'd have to resolve conflicts and you already have approvals. Thanks for making the fix!

Comment (robbie-demuth) [reply pkg/lint/lint.go:33]: Can you also mark that this closes https://github.com/helm/helm/issues/7273?

Comment (mattfarina) [reply pkg/lint/lint.go:33]: @robbie-demuth that is a good questions. Notice that the `Values` function is still present with the same API definition. There is now a new public function named `ValuesWithOverrides` with the new signature. The Go public API change was additive in the form of a new function with the new signature.

**Comment by mattfarina** [comment]
@technosophos this is ready for you to merge when ready.

**Comment by soad666p** [comment]
this is still a problem in helm v3.4.1

**Pull request merged**
